### PR TITLE
Fewer AddressExposedReason::TOO_CONSERVATIVE 

### PR DESCRIPTION
**Pull request by sandreenko**
A few improvements here and there, no actual regressions. The good benefit is that we can tail-call more methods and it should affect the performance better than just the size decrease.

```
Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.windows.x64\coreclr_tests.pmi.windows.x64.checked.mch
 Total bytes of delta: -392 (-2.39% of base)
 85 total files with Code Size differences (80 improved, 5 regressed), 7 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.windows.x64\libraries.crossgen2.windows.x64.checked.mch
 Total bytes of delta: -70 (-2.86% of base)
 10 total files with Code Size differences (10 improved, 0 regressed), 0 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.windows.x64\libraries.pmi.windows.x64.checked.mch
 Total bytes of delta: -69 (-2.33% of base)
 10 total files with Code Size differences (10 improved, 0 regressed), 0 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.windows.x64\libraries_tests.pmi.windows.x64.checked.mch
 Total bytes of delta: -68 (-1.01% of base)
 20 total files with Code Size differences (17 improved, 3 regressed), 1 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.Linux.arm64\coreclr_tests.pmi.Linux.arm64.checked.mch
 Total bytes of delta: -184 (-0.81% of base)
 46 total files with Code Size differences (46 improved, 0 regressed), 5 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.Linux.arm64\libraries.crossgen2.Linux.arm64.checked.mch
 Total bytes of delta: -8 (-0.85% of base)
 2 total files with Code Size differences (2 improved, 0 regressed), 0 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.Linux.arm64\libraries.pmi.Linux.arm64.checked.mch
 Total bytes of delta: -24 (-0.20% of base)
 6 total files with Code Size differences (6 improved, 0 regressed), 1 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.Linux.arm64\libraries_tests.pmi.Linux.arm64.checked.mch
 Total bytes of delta: -56 (-0.34% of base)
 17 total files with Code Size differences (15 improved, 2 regressed), 7 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.Linux.x64\benchmarks.run.Linux.x64.checked.mch
 Total bytes of delta: -17 (-0.56% of base)
 5 total files with Code Size differences (5 improved, 0 regressed), 0 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.Linux.x64\coreclr_tests.pmi.Linux.x64.checked.mch
 Total bytes of delta: -466 (-2.44% of base)
 39 total files with Code Size differences (38 improved, 1 regressed), 8 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.Linux.x64\libraries.crossgen2.Linux.x64.checked.mch
 Total bytes of delta: -16 (-2.36% of base)
 2 total files with Code Size differences (2 improved, 0 regressed), 0 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.Linux.x64\libraries.pmi.Linux.x64.checked.mch
 Total bytes of delta: -34 (-0.60% of base)
 6 total files with Code Size differences (6 improved, 0 regressed), 0 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.Linux.x64\libraries_tests.pmi.Linux.x64.checked.mch
 Total bytes of delta: -216 (-1.30% of base)
 23 total files with Code Size differences (22 improved, 1 regressed), 1 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.windows.arm64\benchmarks.run.windows.arm64.checked.mch
 Total bytes of delta: -20 (-0.72% of base)
 5 total files with Code Size differences (5 improved, 0 regressed), 0 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.windows.arm64\coreclr_tests.pmi.windows.arm64.checked.mch
 Total bytes of delta: -392 (-1.21% of base)
 98 total files with Code Size differences (98 improved, 0 regressed), 11 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.windows.arm64\libraries.crossgen2.windows.arm64.checked.mch
 Total bytes of delta: -44 (-1.14% of base)
 11 total files with Code Size differences (11 improved, 0 regressed), 0 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.windows.arm64\libraries.pmi.windows.arm64.checked.mch
 Total bytes of delta: -48 (-0.32% of base)
 10 total files with Code Size differences (10 improved, 0 regressed), 7 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.windows.arm64\libraries_tests.pmi.windows.arm64.checked.mch
 Total bytes of delta: -80 (-0.51% of base)
 23 total files with Code Size differences (21 improved, 2 regressed), 5 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.windows.x86\benchmarks.run.windows.x86.checked.mch
 Total bytes of delta: -63 (-0.74% of base)
 16 total files with Code Size differences (15 improved, 1 regressed), 2 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.windows.x86\coreclr_tests.pmi.windows.x86.checked.mch
 Total bytes of delta: -400 (-0.45% of base)
 87 total files with Code Size differences (82 improved, 5 regressed), 45 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.windows.x86\libraries.crossgen2.windows.x86.checked.mch
 Total bytes of delta: -140 (-0.41% of base)
 61 total files with Code Size differences (48 improved, 13 regressed), 0 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.windows.x86\libraries.pmi.windows.x86.checked.mch
 Total bytes of delta: -161 (-0.29% of base)
 77 total files with Code Size differences (58 improved, 19 regressed), 7 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.windows.x86\libraries_tests.pmi.windows.x86.checked.mch
 Total bytes of delta: -582 (-0.56% of base)
 187 total files with Code Size differences (173 improved, 14 regressed), 19 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.Linux.arm\coreclr_tests.pmi.Linux.arm.checked.mch
 Total bytes of delta: -74 (-0.42% of base)
 38 total files with Code Size differences (33 improved, 5 regressed), 0 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.Linux.arm\libraries.crossgen2.Linux.arm.checked.mch
 Total bytes of delta: -4 (-0.54% of base)
 2 total files with Code Size differences (2 improved, 0 regressed), 0 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.Linux.arm\libraries.pmi.Linux.arm.checked.mch
 Total bytes of delta: -4 (-0.57% of base)
 2 total files with Code Size differences (2 improved, 0 regressed), 0 unchanged.

Running asm diffs of D:\Sergey\git\runtime\artifacts\spmi\mch\5ed35c58-857b-48dd-a818-7c0136dc9f73.Linux.arm\libraries_tests.pmi.Linux.arm.checked.mch
 Total bytes of delta: -28 (-0.46% of base)
 14 total files with Code Size differences (13 improved, 1 regressed), 1 unchanged.
```

Enreg stat:
```diff
Locals enregistration statistics:
-total number of locals: 346731, number of enregistered: 309775, notEnreg: 36956, ratio: 0.89
+total number of locals: 346731, number of enregistered: 309776, notEnreg: 36955, ratio: 0.89
total number of struct locals: 10920, number of enregistered: 86, notEnreg: 10834, ratio: 0.01
-total number of primitive locals: 335811, number of enregistered: 309689, notEnreg: 26122, ratio: 0.92
+total number of primitive locals: 335811, number of enregistered: 309690, notEnreg: 26121, ratio: 0.92
-m_addrExposed 16176, ratio: 0.44
+m_addrExposed 14059, ratio: 0.38
m_notRegSizeStruct 403, ratio: 0.01
m_localField 3219, ratio: 0.09
+m_VMNeedsStackAddr 2116, ratio: 0.06
                                    
m_liveInOutHndlr 642, ratio: 0.02
m_blockOp 449, ratio: 0.01
m_structArg 26, ratio: 0.00
m_depField 952, ratio: 0.03
m_noRegVars 14819, ratio: 0.40
m_castTakesAddr 12, ratio: 0.00
m_swizzleArg 171, ratio: 0.00
m_blockOpRet 87, ratio: 0.00

Addr exposed details:
-m_parentExposed 5123, ratio: 0.32
+m_parentExposed 5123, ratio: 0.36
-m_tooConservative 3434, ratio: 0.21
+m_tooConservative 235, ratio: 0.02
-m_escapeAddress 7384, ratio: 0.46
+m_escapeAddress 8466, ratio: 0.60
m_wideIndir 7, ratio: 0.00
-m_copyFldByFld 228, ratio: 0.01
+m_copyFldByFld 228, ratio: 0.02
```

Diff example:

Before, tail call is blocked because JIT thinks there is an address exposed var:
```
IN0004: 000031 call     hackishModuleName:hackishMethodName()
IN0005: 000036 nop      

G_M57568_IG03:        ; offs=000037H, size=0005H, bbWeight=1    PerfScore 1.25, epilog, nogc, extend

IN000c: 000037 add      rsp, 40
IN000d: 00003B ret      
```
After, do tailcall.
```
G_M57568_IG03:        ; offs=000028H, size=0005H, bbWeight=1    PerfScore 2.00, epilog, nogc, extend

IN0008: 000028 jmp      hackishModuleName:hackishMethodName()
```

### CODE REVIEW

File: src/coreclr/jit/lower.cpp [hunk 1]
```diff
@@ -4409,8 +4414,12 @@ void Lowering::InsertPInvokeCallEpilog(GenTreeCall* call)
 
         // First argument is the address of the frame variable.
         GenTree* frameAddr =
-            new (comp, GT_LCL_VAR) GenTreeLclVar(GT_LCL_VAR, TYP_BYREF, comp->lvaInlinedPInvokeFrameVar);
-        frameAddr->SetOperRaw(GT_LCL_VAR_ADDR);
+            new (comp, GT_LCL_FLD_ADDR) GenTreeLclFld(GT_LCL_FLD_ADDR, TYP_BYREF, comp->lvaInlinedPInvokeFrameVar, 0);
```
[the remark this answers is not in the document]
Comment (sandreenko) [reply src/coreclr/jit/lower.cpp:4417]: Good catch, no reasons. Fixed.

File: src/coreclr/jit/lclvars.cpp [hunk 2]
```diff
@@ -4348,6 +4347,7 @@ void Compiler::lvaMarkLocalVars()
             lvaPSPSym            = lvaGrabTempWithImplicitUse(false DEBUGARG("PSPSym"));
             LclVarDsc* lclPSPSym = &lvaTable[lvaPSPSym];
             lclPSPSym->lvType    = TYP_I_IMPL;
+            lvaSetVarDoNotEnregister(lvaPSPSym DEBUGARG(DoNotEnregisterReason::VMNeedsStackAddr));
```
Comment (sandreenko) [inline src/coreclr/jit/lclvars.cpp:4350]: now this is the main reason for `VMNeedsStackAddr` but if I understand correctly we actually report this to VM so we can't do anything about it.

File: src/coreclr/jit/lower.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jakobbotsch) [reply src/coreclr/jit/lower.cpp:4417]: Looks like you didn't push this fix yet.

**Comment by jakobbotsch** [comment]
There is an x86 test failure related to varargs it seems, is it unrelated?

**Comment by sandreenko** [comment]
> There is an x86 test failure related to varargs it seems, is it unrelated?

looks related, there are no asm diffs for this method but maybe we tell VM something differently. I will check.

**Comment by sandreenko** [comment]
The failure was fixed and improved the dump a little.

**Review by jakobbotsch — APPROVED** [review]

**Comment by jakobbotsch** [comment]
Failures are #59541

**Pull request merged**